### DISCUSSION TITLE
Turn off "autofix" for a specific rule

### DISCUSSION THREAD

**Discussion by MarkLyck** [biomejs/biome#2104] [Ideas]
BiomeJS seems to have no way to turn off "autofix" for a specific rule. It will either get autofixed if it's "safe" when the file is saved or it won't be linted at all.

This can be very frustrating if you are in the middle of coding e.g. a `let` variable, and hit save before actually using the "let" in your function, in this case I'd prefer the `useConst` rule to throw an error/warning, but not change my `let` into a `const` automatically when I save, so I have to go back and change it to a `let` again and finish writing my function.

This is one example and is obviously different on how people code, but I'm sure this could be useful for other rules as well.

Example of what this could look like in `biome.json`:

```json
"useConst": {
  "level": "error",
  "options": {
    "disableAutoFix": true
  }
}
```

**Comment by ematipico** [comment]
That's because you configured your IDE to apply lint safe fixes on save. You can disable it.

**Reply by MarkLyck** [reply]
@ematipico I'm aware I am applying safe fixes on save. I want to continue doing that in 99% of cases. But there are certain rules that can be frustrating when they auto-fix if you save your files very often.

I want to turn off a specific rule from auto linting my code, not disable safe auto fix on save entirely.

Turning it off entirely and manually running linting is obviously more frustrating than leaving it on and on occasion correcting it. But the developer experience can be improved by rule based options to turn off auto-fix.

**Reply by ematipico** [reply]
>I want to turn off a specific rule from reformatting my code, not disable safe auto fix on save entirely.

Apologies @MarkLyck, but now I'm more confused 😅 you now talk about formatting, but the linter doesn't format any code.

Regardless, I am a bit worried by the retaliation of the initial proposal, because it could give a precedent to **also** change the _safeness_ of the fixes, which is something we shouldn't do.

**Reply by MarkLyck** [reply]
@ematipico sorry, I meant auto-fix of the lint rules, not formatting. (updated my comment above for clarity)

It would not affect the safety of any fixes to turn off auto-fix for a single rule. I would still expect the rule to throw and `error` or `warn` depending on the level it is set to when you run the linting, just not apply the safe changes automatically on save for the specific rule I opted out of.

It should of course be an `opt-out` option, not on by default. It just allows a better developer experience for people who save their files very often with unfinished code.

This was a feature of ESLint as well, either through the [eslint-plugin-no-autofix](https://www.npmjs.com/package/eslint-plugin-no-autofix) plugin, which allowed disabling autofix for specific rules. This extension has 34k weekly downloads, 

There's another plugin as well that does the same: https://www.npmjs.com/package/eslint-plugin-disable-autofix with 29k weekly downloads.

So I don't think I'm the only developer who wants to accomplish this.

The eslint VScode extension also allows disabling autofix for specific rules like this: (no plugin needed)

```json
"eslint.codeActionsOnSave.rules": [
        "!prefer-const",
        "*"
    ],
```

the above would disable auto fix on save for the `prefer-const` eslint rule. 

I would love to see feature parity for with this for biomeJS.

**Comment by Conaclos** [comment]
Closing because we have added this feature on main. It will be available in Biome 1.8.

**Reply by aleclarson** [reply]
For anyone wondering, here's how you configure it:

```json
"linter": {
  "enabled": true,
  "rules": {
    "style": {
      "useConst": {
        "fix": "none",
        "level": "warn"
      }
    }
  }
}
```

**Reply by lukepighetti** [reply]
<img width="612" height="483" alt="Screenshot 2025-11-12 at 4 12 46 PM" src="https://github.com/user-attachments/assets/cd5f899c-1fe4-4b61-a138-f4f36dd914ee" />

this doesn't seem to work with schema version `2.3.4`

**Reply by vladyslav-n** [reply]
> lukepighetti
The [schema](https://biomejs.dev/schemas/2.3.4/schema.json) is quite strict – it just doesn't provide this option for **all** the rules – just for some of them. You can search for `FixKind` and see which rules support it.
